toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
-      builtin_ir, tokenized_source, parse_tree, consumer, vlog_stream_);
+      builtin_ir, tokenized_source, parse_tree, *consumer, vlog_stream_);
   has_errors |= semantics_ir.has_errors();
   CARBON_VLOG() << "*** SemanticsIR::MakeFromParseTree done ***\n";
-  if (dump_mode == DumpMode::SemanticsIR) {
-    consumer.Flush();
-    semantics_ir.Print(output_stream_, semantics_ir_include_builtins);
-    return !has_errors;
+  if (options.dump_semantics_ir) {
+    consumer->Flush();
+    semantics_ir.Print(output_stream_, options.builtin_semantics_ir);
   }
   CARBON_VLOG() << "semantics_ir: " << semantics_ir;
+  if (options.phase == Phase::Syntax) {
+    return !has_errors;
+  }
 
   // Unlike previous steps, errors block further progress.
   if (has_errors) {
-    CARBON_VLOG() << "Unable to dump llvm-ir due to prior errors.";
+    CARBON_VLOG() << "*** Stopping before lowering due to syntax errors ***";
     return false;
   }
+  consumer->Flush();
 
   CARBON_VLOG() << "*** LowerToLLVM ***\n";
   llvm::LLVMContext llvm_context;
-  const std::unique_ptr<llvm::Module> module =
-      LowerToLLVM(llvm_context, input_file_name, semantics_ir, vlog_stream_);
+  const std::unique_ptr<llvm::Module> module = LowerToLLVM(
+      llvm_context, options.input_file_name, semantics_ir, vlog_stream_);
   CARBON_VLOG() << "*** LowerToLLVM done ***\n";
-  if (dump_mode == DumpMode::LLVMIR) {
-    consumer.Flush();
+  if (options.dump_llvm_ir) {
     module->print(output_stream_, /*AAW=*/nullptr,
                   /*ShouldPreserveUseListOrder=*/true);
-    return !has_errors;
   }
   if (vlog_stream_) {
     CARBON_VLOG() << "module: ";
     module->print(*vlog_stream_, /*AAW=*/nullptr,
                   /*ShouldPreserveUseListOrder=*/false,
                   /*IsForDebug=*/true);
   }
+  if (options.phase == Phase::Lower) {
+    return true;
+  }
 
-  if (dump_mode == DumpMode::Assembly) {
-    consumer.Flush();
-    CodeGen codegen(*module, target_triple, error_stream_, output_stream_);
-    has_errors |= !codegen.PrintAssembly();
-    return !has_errors;
+  CARBON_VLOG() << "*** CodeGen ***\n";
+  std::optional<CodeGen> codegen =
+      CodeGen::Create(*module, options.target, error_stream_);
+  if (!codegen) {
+    return false;
+  }
+  if (vlog_stream_) {
+    CARBON_VLOG() << "assembly:\n";
+    codegen->EmitAssembly(*vlog_stream_);
   }
 
-  if (dump_mode == DumpMode::ObjectCode) {
+  if (options.output_file_name == "-") {
+    if (options.force_obj_output) {
+      if (!codegen->EmitObject(output_stream_)) {
+        return false;
+      }
+    } else {
+      if (!codegen->EmitAssembly(output_stream_)) {
+        return false;
+      }
+    }
+  } else {
+    llvm::SmallString<256> output_file_name = options.output_file_name;
+    if (output_file_name.empty()) {
+      output_file_name = options.input_file_name;
+      llvm::sys::path::replace_extension(output_file_name,
+                                         options.asm_output ? ".s" : ".o");
```

Comment:
if the output file is empty and the codegen flag is set to assembly why are we creating an output file instead of printing the assembly directly to the stdout?

Reply:
I think I covered this somewhat with the answer above, but just to be clear -- this is a fairly strong convention in compilers and several other command line tools to synthesize a filename with the correct extension from the input filename.

I think outside of compiler test suites, this is also the likely desired behavior.